plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/bitbucket.ts

```diff
@@ -0,0 +1,251 @@
+    body: JSON.stringify({
+      scm: 'git',
+      description: description,
+      is_private: repoVisibility === 'private',
+    }),
+    headers: {
+      Authorization: authorization,
+      'Content-Type': 'application/json',
+    },
+  };
+
+  let response: Response;
+  try {
+    response = await fetch(
+      `https://api.bitbucket.org/2.0/repositories/${owner}/${repo}`,
+      options,
+    );
+  } catch (e) {
+    throw new Error(`Unable to create repository, ${e}`);
+  }
+
+  if (response.status !== 200) {
+    throw new Error(
+      `Unable to create repository, ${response.status} ${
+        response.statusText
+      }, ${await response.text()}`,
+    );
+  }
+
+  const r = await response.json();
+  let remoteUrl = '';
+  for (const link of r.links.clone) {
+    if (link.name === 'https') {
+      remoteUrl = link.href;
+    }
+  }
+
+  // TODO use the urlReader to get the default branch
+  const repoContentsUrl = `${r.links.html.href}/src/master`;
+  return { remoteUrl, repoContentsUrl };
+};
+
+const createBitbucketServerRepository = async (opts: {
+  host: string;
+  owner: string;
+  repo: string;
+  description: string;
+  repoVisibility: 'private' | 'public';
+  authorization: string;
+}) => {
+  const {
+    host,
+    owner,
+    repo,
+    description,
+    authorization,
+    repoVisibility,
+  } = opts;
+
+  let response: Response;
+  const options: RequestInit = {
+    method: 'POST',
+    body: JSON.stringify({
+      name: repo,
+      description: description,
+      is_private: repoVisibility === 'private',
+    }),
+    headers: {
+      Authorization: authorization,
+      'Content-Type': 'application/json',
+    },
+  };
+
+  try {
+    response = await fetch(
+      `https://${host}/rest/api/1.0/projects/${owner}/repos`,
+      options,
+    );
+  } catch (e) {
+    throw new Error(`Unable to create repository, ${e}`);
+  }
+
+  if (response.status !== 201) {
+    throw new Error(
+      `Unable to create repository, ${response.status} ${
+        response.statusText
+      }, ${await response.text()}`,
+    );
+  }
+
+  const r = await response.json();
+  let remoteUrl = '';
+  for (const link of r.links.clone) {
+    if (link.name === 'http') {
+      remoteUrl = link.href;
+    }
+  }
+
+  const repoContentsUrl = `${r.links.self[0].href}`;
+  return { remoteUrl, repoContentsUrl };
+};
+
+const getAuthorizationHeader = (config: BitbucketIntegrationConfig) => {
+  if (config.username && config.appPassword) {
+    const buffer = Buffer.from(
+      `${config.username}:${config.appPassword}`,
+      'utf8',
+    );
+
+    return `Basic ${buffer.toString('base64')}`;
+  }
+
+  if (config.token) {
+    return `Bearer ${config.token}`;
+  }
+
+  throw new Error(
+    `Authorization has not been provided for Bitbucket. Please add either username + appPassword or token to the Integrations config`,
+  );
+};
+
+export function createPublishBitbucketAction(options: {
+  integrations: ScmIntegrations;
```

Comment:
general note - `ScmIntegrationRegistry` is the type, `ScmIntegrations` is the concrete class - you may want to search and replace some of this